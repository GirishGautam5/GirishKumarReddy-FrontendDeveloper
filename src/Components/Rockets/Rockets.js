import React from "react";

export const handleFilterChange = (event, setFilter, filter) => {
  const { name, value } = event.target;
  setFilter({ ...filter, [name]: value });
};

export  const filteredData =(rockets, filter)=> rockets?.filter((item) => {
  return (
    item?.rocket_name?.includes(filter.name) &&
    item?.active?.toString()?.includes(filter.status) &&
    item?.cost_per_launch?.toString()?.includes(filter.cost)
  );
});
export default function Rockets() {
  const [rockets, setRockets] = React.useState(null);
  const [loading, setLoading] = React.useState(true);

  const [filter, setFilter] = React.useState({ name: "", status: "", cost: "" });

  React.useEffect(() => {
    const fetchRockets = async () => {
      const res = await fetch(`https://api.spacexdata.com/v3/rockets`);
      const data = await res.json();
      setRockets(data);
      setLoading(false);
    };

    fetchRockets();
  }, []);

  

 
const rocketsDataAfterFilter = filteredData(rockets, filter)
  const showFilteredData = filter.name || filter.status || filter.cost;
  const rocketsData = showFilteredData ? rocketsDataAfterFilter : rockets;

  return loading ? (
    <div className="loader"></div>
  ) : (
    <section className="rocket">
      <div className="overlay">
        <h1 className="heading">Rockets</h1>

        <div className="max-width grid grid-cols-1 gap-5 md:grid-cols-2 lg:grid-cols-3 mt-10">
          <label className="text-white">
            Rocket Name:{" "}
            <input
              type="text"
              name="name"
              placeholder="Search by Rocket name"
              value={filter.name}
              onChange={(event)=>handleFilterChange(event, setFilter,filter)}
            />
          </label>
          <label className="text-white">
            Rocket Status:{" "}
            <input
              type="text"
              name="status"
              placeholder="Search by status(true or false)"
              value={filter.status}
              onChange={(event)=>handleFilterChange(event, setFilter,filter)}
            />
          </label>
          <label className="text-white">
            Total Cost:{" "}
            <input
              type="text"
              name="cost"
              placeholder="Search by total cost"
              value={filter.cost}
              onChange={(event)=>handleFilterChange(event, setFilter,filter)}
            />
          </label>
          {rocketsData?.map((rocket) => (
            <article key={rocket?.id} className="articles">
              <img
                src={rocket?.flickr_images[0]}
                alt={rocket?.rocket_name}
                className="h-72 w-full object-cover"
              />
              <div className="p-5">
                <h2 className="text-white font-bold text-xl my-1">
                  {rocket?.rocket_name}
                </h2>
                <h3
                  className={`capitalize ${
                    rocket?.active === true ? "text-green-500" : "text-red-500"
                  }`}
                >
                  Status: {rocket?.active === true ? "true" : "false"}
                </h3>
                <p>Cost per launch: {rocket?.cost_per_launch}</p>
                <p className="text-white opacity-75 text-sm mb-5">{`${rocket?.description.substring(
                  0,
                  100
                )}...`}</p>
                <button className="btn">Learn more</button>
              </div>
            </article>
          ))}
          {showFilteredData && rocketsDataAfterFilter?.length === 0 && (
            <h1 className="noResult">No Search Results ....</h1>
          )}
        </div>
      </div>
    </section>
  );
}
