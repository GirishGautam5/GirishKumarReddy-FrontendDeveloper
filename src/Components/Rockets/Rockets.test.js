import { shallow } from "enzyme";
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Rockets, { filteredData, handleFilterChange } from "./Rockets";
import React from "react";
import toJson from "enzyme-to-json";

Enzyme.configure({ adapter: new Adapter() });

describe("Rockets test suite", () => {
  const rocketComponent = shallow(<Rockets />);
  const useEffectMock = jest.spyOn(React, "useEffect");
  beforeEach(() => {
    useEffectMock.mockClear();
  });
  it("Rocket component  test case", () => {
    const rocketComponent = shallow(<Rockets />);
    expect(rocketComponent).toMatchSnapshot();
  });
  it("Rocket component test for useffect", async () => {
    const rocketsData = [
      {
        id: "1",
        rocket_name: "Falcon 9",
        active: false,
        description: "This is Falcon 9 Rocket",
        flickr_images: [{}],
      },
    ];

    const mockUseEffect = () => {
      useEffectMock.mockImplementationOnce((f) => f());
    };
    mockUseEffect();
    mockUseEffect();
    mockUseEffect();
    mockUseEffect();
    React.useState = jest
      .fn()
      .mockReturnValueOnce([rocketsData, (val) => val])
      .mockReturnValueOnce([false, (val) => val])
      .mockReturnValueOnce([{ name: "" }, (val) => val]);
    const rocketComponent = shallow(<Rockets />);
    expect(rocketComponent).toBeDefined();
    expect(toJson(rocketComponent)).toMatchSnapshot();
  });
  it("Rocket component test for rocket status true", () => {
    const rocketsData = [
      {
        id: "1",
        rocket_name: "Falcon 9",
        active: true,
        description: "This is Falcon 9 Rocket",
        flickr_images: [{}],
      },
    ];

    React.useState = jest
      .fn()
      .mockReturnValueOnce([rocketsData, (val) => val])
      .mockReturnValueOnce([false, (val) => val])
      .mockReturnValueOnce([{ name: "" }, (val) => val]);
    const rocketComponent = shallow(<Rockets />);
    expect(rocketComponent).toBeDefined();
  });
  it("Rocket component test for search value", async () => {
    const rocketsData = [
      {
        id: "1",
        rocket_name: "Falcon 9",
        active: true,
        description: "This is Falcon 9 Rocket",
        flickr_images: [{}],
        cost_per_launch: 9000000,
      },
    ];

    const mockUseEffect = () => {
      useEffectMock.mockImplementationOnce((f) => f());
    };
    mockUseEffect();
    mockUseEffect();
    mockUseEffect();
    mockUseEffect();
    React.useState = jest
      .fn()
      .mockReturnValueOnce([rocketsData, (val) => val])
      .mockReturnValueOnce([false, (val) => val])
      .mockReturnValueOnce([{ name: "Falcon 9" }, (val) => val]);
    const rocketComponent = shallow(<Rockets />);
    expect(rocketComponent).toBeDefined();
  });
  it("handleFilterChange testcase", () => {
    const event = {
      target: {
        name: "status",
        value: "true",
      },
    };
    const setFilter = jest.fn();
    const filter = {
      name: "",
      status: "",
      cost: "",
    };
    handleFilterChange(event, setFilter, filter);
  });
  it("filtereredData test case", () => {
    const rockets = [
      {
        id: "1",
        rocket_name: "Falcon 9",
        active: true,
        description: "This is Falcon 9 Rocket",
        flickr_images: [{}],
        cost_per_launch: 9000000,
      },
    ];
    const filter = {
      name: "Falcon 9",
      status: true,
      cost: 9000000,
    };
    const result = filteredData(rockets, filter);
    expect(result).toEqual(rockets);
  });
  it("test case for input onChange", () => {
    rocketComponent.find("input").forEach((textInput) => {
      textInput.props().onChange();
    });
  });
});
