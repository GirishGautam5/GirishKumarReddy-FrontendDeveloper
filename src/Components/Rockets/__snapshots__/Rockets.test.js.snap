// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets test suite Rocket component  test case 1`] = `ShallowWrapper {}`;

exports[`Rockets test suite Rocket component test for useffect 1`] = `
<section
  className="rocket"
>
  <div
    className="overlay"
  >
    <h1
      className="heading"
    >
      Rockets
    </h1>
    <div
      className="max-width grid grid-cols-1 gap-5 md:grid-cols-2 lg:grid-cols-3 mt-10"
    >
      <label
        className="text-white"
      >
        Rocket Name:
         
        <input
          name="name"
          onChange={[Function]}
          placeholder="Search by Rocket name"
          type="text"
          value=""
        />
      </label>
      <label
        className="text-white"
      >
        Rocket Status:
         
        <input
          name="status"
          onChange={[Function]}
          placeholder="Search by status(true or false)"
          type="text"
        />
      </label>
      <label
        className="text-white"
      >
        Total Cost:
         
        <input
          name="cost"
          onChange={[Function]}
          placeholder="Search by total cost"
          type="text"
        />
      </label>
      <article
        className="articles"
        key="1"
      >
        <img
          alt="Falcon 9"
          className="h-72 w-full object-cover"
          src={Object {}}
        />
        <div
          className="p-5"
        >
          <h2
            className="text-white font-bold text-xl my-1"
          >
            Falcon 9
          </h2>
          <h3
            className="capitalize text-red-500"
          >
            Status: 
            false
          </h3>
          <p>
            Cost per launch: 
          </p>
          <p
            className="text-white opacity-75 text-sm mb-5"
          >
            This is Falcon 9 Rocket...
          </p>
          <button
            className="btn"
          >
            Learn more
          </button>
        </div>
      </article>
    </div>
  </div>
</section>
`;
